File: polyglot/runtime_template_resolver/py/examples/basic_usage.py
#!/usr/bin/env python3
"""
Runtime Template Resolver - Basic Usage Examples

This script demonstrates core features of the runtime_template_resolver package:
- Simple placeholder resolution
- Nested path resolution
- Default values
- Missing value strategies
- Object/config resolution
- Template compilation
- Placeholder extraction and validation
"""
import sys
import os

# Add parent src to path for local development
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from runtime_template_resolver import (
    TemplateResolver,
    resolve,
    resolve_many,
    resolve_object,
    validate,
    extract,
    compile,
    MissingStrategy,
    ResolverOptions,
    SecurityError,
    ValidationError,
    MissingValueError,
)


# =============================================================================
# Example 1: Simple Placeholder Resolution
# =============================================================================
def example1_simple_resolution() -> None:
    """
    Demonstrates basic placeholder resolution with simple key-value context.
    """
    print("\n" + "=" * 60)
    print("Example 1: Simple Placeholder Resolution")
    print("=" * 60)

    resolver = TemplateResolver()

    # Simple greeting template
    template = "Hello, {{name}}! Welcome to {{company}}."
    context = {"name": "Alice", "company": "ACME Corp"}

    result = resolver.resolve(template, context)
    print(f"Template: {template}")
    print(f"Context:  {context}")
    print(f"Result:   {result}")


# =============================================================================
# Example 2: Nested Path Resolution
# =============================================================================
def example2_nested_paths() -> None:
    """
    Demonstrates resolution of nested object paths using dot notation.
    """
    print("\n" + "=" * 60)
    print("Example 2: Nested Path Resolution")
    print("=" * 60)

    resolver = TemplateResolver()

    template = "User {{user.profile.name}} ({{user.profile.email}}) has role: {{user.role}}"
    context = {
        "user": {
            "profile": {
                "name": "Bob Smith",
                "email": "bob@example.com"
            },
            "role": "admin"
        }
    }

    result = resolver.resolve(template, context)
    print(f"Template: {template}")
    print(f"Result:   {result}")


# =============================================================================
# Example 3: Array Access
# =============================================================================
def example3_array_access() -> None:
    """
    Demonstrates resolution of array elements using bracket notation.
    """
    print("\n" + "=" * 60)
    print("Example 3: Array Access")
    print("=" * 60)

    resolver = TemplateResolver()

    template = "Top items: {{items[0]}}, {{items[1]}}, {{items[2]}}"
    context = {
        "items": ["Apple", "Banana", "Cherry", "Date"]
    }

    result = resolver.resolve(template, context)
    print(f"Template: {template}")
    print(f"Result:   {result}")


# =============================================================================
# Example 4: Default Values
# =============================================================================
def example4_default_values() -> None:
    """
    Demonstrates using default values for missing placeholders.
    """
    print("\n" + "=" * 60)
    print("Example 4: Default Values")
    print("=" * 60)

    resolver = TemplateResolver()

    # Various default value syntaxes
    templates = [
        ('DB Host: {{db.host | "localhost"}}', {}),
        ("DB Port: {{db.port | '5432'}}", {}),
        ("Environment: {{env | production}}", {}),
        ('API Key: {{api_key | "not-set"}}', {"api_key": "secret-123"}),
    ]

    for template, context in templates:
        result = resolver.resolve(template, context)
        print(f"Template: {template}")
        print(f"Result:   {result}")
        print()


# =============================================================================
# Example 5: Missing Value Strategies
# =============================================================================
def example5_missing_strategies() -> None:
    """
    Demonstrates different strategies for handling missing values.
    """
    print("\n" + "=" * 60)
    print("Example 5: Missing Value Strategies")
    print("=" * 60)

    resolver = TemplateResolver()
    template = "Value: {{missing_key}}"
    context = {}

    strategies = [
        (MissingStrategy.EMPTY, "Replace with empty string"),
        (MissingStrategy.KEEP, "Keep original placeholder"),
        (MissingStrategy.DEFAULT, "Use default value (empty if none)"),
    ]

    for strategy, description in strategies:
        opts = ResolverOptions(missing_strategy=strategy)
        result = resolver.resolve(template, context, options=opts)
        print(f"Strategy {strategy.value}: {description}")
        print(f"  Result: '{result}'")
        print()

    # ERROR strategy throws exception
    print("Strategy ERROR: Throws exception")
    try:
        opts = ResolverOptions(missing_strategy=MissingStrategy.ERROR, throw_on_error=True)
        resolver.resolve(template, context, options=opts)
    except MissingValueError as e:
        print(f"  Caught: {e}")


# =============================================================================
# Example 6: Resolve Object (Config Templates)
# =============================================================================
def example6_resolve_object() -> None:
    """
    Demonstrates resolving templates within nested objects/configs.
    """
    print("\n" + "=" * 60)
    print("Example 6: Resolve Object (Config Templates)")
    print("=" * 60)

    resolver = TemplateResolver()

    # Simulated config object with templates
    config = {
        "database": {
            "url": "postgres://{{db.host}}:{{db.port}}/{{db.name}}",
            "pool_size": 10
        },
        "api": {
            "base_url": "https://{{api.domain}}/v{{api.version}}",
            "endpoints": [
                "/users/{{user_id}}",
                "/orders/{{order_id}}"
            ]
        },
        "messages": {
            "welcome": "Welcome {{user.name}} to {{app.name}}!"
        }
    }

    context = {
        "db": {"host": "localhost", "port": "5432", "name": "myapp"},
        "api": {"domain": "api.example.com", "version": "2"},
        "user_id": "123",
        "order_id": "456",
        "user": {"name": "Alice"},
        "app": {"name": "MyApp"}
    }

    resolved = resolver.resolve_object(config, context)

    print("Original config with templates:")
    print(f"  database.url: {config['database']['url']}")
    print(f"  api.base_url: {config['api']['base_url']}")
    print()
    print("Resolved config:")
    print(f"  database.url: {resolved['database']['url']}")
    print(f"  api.base_url: {resolved['api']['base_url']}")
    print(f"  api.endpoints: {resolved['api']['endpoints']}")
    print(f"  messages.welcome: {resolved['messages']['welcome']}")


# =============================================================================
# Example 7: Template Compilation
# =============================================================================
def example7_compilation() -> None:
    """
    Demonstrates compiling templates for repeated use.
    """
    print("\n" + "=" * 60)
    print("Example 7: Template Compilation")
    print("=" * 60)

    # Compile template once
    email_template = compile(
        "Dear {{name}},\n\n"
        "Your order #{{order_id}} has been {{status}}.\n\n"
        "Best regards,\n{{company}}"
    )

    # Use multiple times with different contexts
    orders = [
        {"name": "Alice", "order_id": "1001", "status": "shipped", "company": "ACME"},
        {"name": "Bob", "order_id": "1002", "status": "delivered", "company": "ACME"},
        {"name": "Charlie", "order_id": "1003", "status": "processing", "company": "ACME"},
    ]

    for order in orders:
        result = email_template(order)
        print(f"--- Order #{order['order_id']} ---")
        print(result)
        print()


# =============================================================================
# Example 8: Placeholder Extraction
# =============================================================================
def example8_extraction() -> None:
    """
    Demonstrates extracting placeholder keys from templates.
    """
    print("\n" + "=" * 60)
    print("Example 8: Placeholder Extraction")
    print("=" * 60)

    templates = [
        "Hello {{name}}!",
        "{{user.profile.name}} works at {{company}}",
        "Items: {{items[0]}}, {{items[1]}}",
        '{{value | "default"}}',
    ]

    for template in templates:
        placeholders = extract(template)
        print(f"Template: {template}")
        print(f"  Placeholders: {placeholders}")
        print()


# =============================================================================
# Example 9: Validation
# =============================================================================
def example9_validation() -> None:
    """
    Demonstrates template validation before resolution.
    """
    print("\n" + "=" * 60)
    print("Example 9: Validation")
    print("=" * 60)

    valid_templates = [
        "{{name}}",
        "{{user.profile.name}}",
        "{{items[0]}}",
        '{{value | "default"}}',
    ]

    invalid_templates = [
        "{{foo@bar}}",  # Invalid character @
        "{{foo..bar}}",  # Empty segment
    ]

    print("Valid templates:")
    for template in valid_templates:
        try:
            validate(template)
            print(f"  {template}")
        except ValidationError as e:
            print(f"  {template} - ERROR: {e}")

    print("\nInvalid templates:")
    for template in invalid_templates:
        try:
            validate(template)
            print(f"  {template} - OK (unexpected)")
        except ValidationError as e:
            print(f"  {template}")
            print(f"    Error: {e}")


# =============================================================================
# Example 10: Security - Private Attribute Protection
# =============================================================================
def example10_security() -> None:
    """
    Demonstrates security protection against private attribute access.
    """
    print("\n" + "=" * 60)
    print("Example 10: Security - Private Attribute Protection")
    print("=" * 60)

    resolver = TemplateResolver()

    dangerous_templates = [
        "{{_private}}",
        "{{__class__}}",
        "{{obj.__dict__}}",
    ]

    print("Attempting to access private attributes:")
    for template in dangerous_templates:
        try:
            opts = ResolverOptions(throw_on_error=True)
            result = resolver.resolve(template, {"_private": "secret"}, options=opts)
            print(f"  {template} -> {result}")
        except SecurityError as e:
            print(f"  {template}")
            print(f"    Blocked: {e}")


# =============================================================================
# Example 11: SDK Convenience Functions
# =============================================================================
def example11_sdk_functions() -> None:
    """
    Demonstrates using SDK-level convenience functions.
    """
    print("\n" + "=" * 60)
    print("Example 11: SDK Convenience Functions")
    print("=" * 60)

    context = {"name": "World", "count": 42}

    # resolve() - single template
    result = resolve("Hello {{name}}!", context)
    print(f"resolve(): {result}")

    # resolve_many() - multiple templates
    results = resolve_many([
        "Hello {{name}}",
        "Count: {{count}}",
        "Goodbye {{name}}"
    ], context)
    print(f"resolve_many(): {results}")

    # resolve_object() - nested objects
    obj = {"greeting": "Hello {{name}}", "data": {"value": "{{count}}"}}
    resolved = resolve_object(obj, context)
    print(f"resolve_object(): {resolved}")


# =============================================================================
# Main Runner
# =============================================================================
def main() -> None:
    """Run all examples sequentially."""
    print("=" * 60)
    print("Runtime Template Resolver - Basic Usage Examples")
    print("=" * 60)

    example1_simple_resolution()
    example2_nested_paths()
    example3_array_access()
    example4_default_values()
    example5_missing_strategies()
    example6_resolve_object()
    example7_compilation()
    example8_extraction()
    example9_validation()
    example10_security()
    example11_sdk_functions()

    print("\n" + "=" * 60)
    print("All examples completed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: polyglot/runtime_template_resolver/mjs/examples/basic-usage.ts
#!/usr/bin/env npx tsx
/**
 * Runtime Template Resolver - Basic Usage Examples
 *
 * This script demonstrates core features of the runtime_template_resolver package:
 * - Simple placeholder resolution
 * - Nested path resolution
 * - Default values
 * - Missing value strategies
 * - Object/config resolution
 * - Template compilation
 * - Placeholder extraction and validation
 */

import {
    TemplateResolver,
    SDK,
    MissingStrategy,
    SecurityError,
    ValidationError,
    MissingValueError,
    validatePlaceholder,
    extractPlaceholders,
} from '../src/index.js';

// =============================================================================
// Example 1: Simple Placeholder Resolution
// =============================================================================
function example1_simpleResolution(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 1: Simple Placeholder Resolution');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();

    const template = 'Hello, {{name}}! Welcome to {{company}}.';
    const context = { name: 'Alice', company: 'ACME Corp' };

    const result = resolver.resolve(template, context);
    console.log(`Template: ${template}`);
    console.log(`Context:  ${JSON.stringify(context)}`);
    console.log(`Result:   ${result}`);
}

// =============================================================================
// Example 2: Nested Path Resolution
// =============================================================================
function example2_nestedPaths(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 2: Nested Path Resolution');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();

    const template = 'User {{user.profile.name}} ({{user.profile.email}}) has role: {{user.role}}';
    const context = {
        user: {
            profile: {
                name: 'Bob Smith',
                email: 'bob@example.com'
            },
            role: 'admin'
        }
    };

    const result = resolver.resolve(template, context);
    console.log(`Template: ${template}`);
    console.log(`Result:   ${result}`);
}

// =============================================================================
// Example 3: Array Access
// =============================================================================
function example3_arrayAccess(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 3: Array Access');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();

    const template = 'Top items: {{items[0]}}, {{items[1]}}, {{items[2]}}';
    const context = {
        items: ['Apple', 'Banana', 'Cherry', 'Date']
    };

    const result = resolver.resolve(template, context);
    console.log(`Template: ${template}`);
    console.log(`Result:   ${result}`);
}

// =============================================================================
// Example 4: Default Values
// =============================================================================
function example4_defaultValues(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 4: Default Values');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();

    const templates: Array<[string, Record<string, unknown>]> = [
        ['DB Host: {{db.host | "localhost"}}', {}],
        ["DB Port: {{db.port | '5432'}}", {}],
        ['Environment: {{env | production}}', {}],
        ['API Key: {{api_key | "not-set"}}', { api_key: 'secret-123' }],
    ];

    for (const [template, context] of templates) {
        const result = resolver.resolve(template, context);
        console.log(`Template: ${template}`);
        console.log(`Result:   ${result}`);
        console.log();
    }
}

// =============================================================================
// Example 5: Missing Value Strategies
// =============================================================================
function example5_missingStrategies(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 5: Missing Value Strategies');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();
    const template = 'Value: {{missing_key}}';
    const context = {};

    const strategies: Array<[MissingStrategy, string]> = [
        [MissingStrategy.EMPTY, 'Replace with empty string'],
        [MissingStrategy.KEEP, 'Keep original placeholder'],
        [MissingStrategy.DEFAULT, 'Use default value (empty if none)'],
    ];

    for (const [strategy, description] of strategies) {
        const result = resolver.resolve(template, context, { missingStrategy: strategy });
        console.log(`Strategy ${strategy}: ${description}`);
        console.log(`  Result: '${result}'`);
        console.log();
    }

    // ERROR strategy throws exception
    console.log('Strategy ERROR: Throws exception');
    try {
        resolver.resolve(template, context, {
            missingStrategy: MissingStrategy.ERROR,
            throwOnError: true
        });
    } catch (e) {
        if (e instanceof MissingValueError) {
            console.log(`  Caught: ${e.message}`);
        }
    }
}

// =============================================================================
// Example 6: Resolve Object (Config Templates)
// =============================================================================
function example6_resolveObject(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 6: Resolve Object (Config Templates)');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();

    const config = {
        database: {
            url: 'postgres://{{db.host}}:{{db.port}}/{{db.name}}',
            poolSize: 10
        },
        api: {
            baseUrl: 'https://{{api.domain}}/v{{api.version}}',
            endpoints: [
                '/users/{{user_id}}',
                '/orders/{{order_id}}'
            ]
        },
        messages: {
            welcome: 'Welcome {{user.name}} to {{app.name}}!'
        }
    };

    const context = {
        db: { host: 'localhost', port: '5432', name: 'myapp' },
        api: { domain: 'api.example.com', version: '2' },
        user_id: '123',
        order_id: '456',
        user: { name: 'Alice' },
        app: { name: 'MyApp' }
    };

    const resolved = resolver.resolveObject(config, context) as typeof config;

    console.log('Original config with templates:');
    console.log(`  database.url: ${config.database.url}`);
    console.log(`  api.baseUrl: ${config.api.baseUrl}`);
    console.log();
    console.log('Resolved config:');
    console.log(`  database.url: ${resolved.database.url}`);
    console.log(`  api.baseUrl: ${resolved.api.baseUrl}`);
    console.log(`  api.endpoints: ${JSON.stringify(resolved.api.endpoints)}`);
    console.log(`  messages.welcome: ${resolved.messages.welcome}`);
}

// =============================================================================
// Example 7: Template Compilation
// =============================================================================
function example7_compilation(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 7: Template Compilation');
    console.log('='.repeat(60));

    // Compile template once
    const emailTemplate = SDK.compile(
        'Dear {{name}},\n\n' +
        'Your order #{{order_id}} has been {{status}}.\n\n' +
        'Best regards,\n{{company}}'
    );

    // Use multiple times with different contexts
    const orders = [
        { name: 'Alice', order_id: '1001', status: 'shipped', company: 'ACME' },
        { name: 'Bob', order_id: '1002', status: 'delivered', company: 'ACME' },
        { name: 'Charlie', order_id: '1003', status: 'processing', company: 'ACME' },
    ];

    for (const order of orders) {
        const result = emailTemplate(order);
        console.log(`--- Order #${order.order_id} ---`);
        console.log(result);
        console.log();
    }
}

// =============================================================================
// Example 8: Placeholder Extraction
// =============================================================================
function example8_extraction(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 8: Placeholder Extraction');
    console.log('='.repeat(60));

    const templates = [
        'Hello {{name}}!',
        '{{user.profile.name}} works at {{company}}',
        'Items: {{items[0]}}, {{items[1]}}',
        '{{value | "default"}}',
    ];

    for (const template of templates) {
        const placeholders = SDK.extract(template);
        console.log(`Template: ${template}`);
        console.log(`  Placeholders: ${JSON.stringify(placeholders)}`);
        console.log();
    }
}

// =============================================================================
// Example 9: Validation
// =============================================================================
function example9_validation(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 9: Validation');
    console.log('='.repeat(60));

    const validTemplates = [
        '{{name}}',
        '{{user.profile.name}}',
        '{{items[0]}}',
        '{{value | "default"}}',
    ];

    const invalidTemplates = [
        '{{foo@bar}}',  // Invalid character @
        '{{foo..bar}}',  // Empty segment
    ];

    console.log('Valid templates:');
    for (const template of validTemplates) {
        try {
            SDK.validate(template);
            console.log(`  ${template}`);
        } catch (e) {
            if (e instanceof ValidationError) {
                console.log(`  ${template} - ERROR: ${e.message}`);
            }
        }
    }

    console.log('\nInvalid templates:');
    for (const template of invalidTemplates) {
        try {
            SDK.validate(template);
            console.log(`  ${template} - OK (unexpected)`);
        } catch (e) {
            if (e instanceof ValidationError) {
                console.log(`  ${template}`);
                console.log(`    Error: ${e.message}`);
            }
        }
    }
}

// =============================================================================
// Example 10: Security - Private Attribute Protection
// =============================================================================
function example10_security(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 10: Security - Private Attribute Protection');
    console.log('='.repeat(60));

    const resolver = new TemplateResolver();

    const dangerousTemplates = [
        '{{_private}}',
        '{{__proto__}}',
        '{{obj.__dict__}}',
    ];

    console.log('Attempting to access private attributes:');
    for (const template of dangerousTemplates) {
        try {
            const result = resolver.resolve(template, { _private: 'secret' }, { throwOnError: true });
            console.log(`  ${template} -> ${result}`);
        } catch (e) {
            if (e instanceof SecurityError) {
                console.log(`  ${template}`);
                console.log(`    Blocked: ${e.message}`);
            }
        }
    }
}

// =============================================================================
// Example 11: SDK Convenience Functions
// =============================================================================
function example11_sdkFunctions(): void {
    console.log('\n' + '='.repeat(60));
    console.log('Example 11: SDK Convenience Functions');
    console.log('='.repeat(60));

    const context = { name: 'World', count: 42 };

    // resolve() - single template
    const result = SDK.resolve('Hello {{name}}!', context);
    console.log(`resolve(): ${result}`);

    // resolveMany() - multiple templates
    const results = SDK.resolveMany([
        'Hello {{name}}',
        'Count: {{count}}',
        'Goodbye {{name}}'
    ], context);
    console.log(`resolveMany(): ${JSON.stringify(results)}`);

    // resolveObject() - nested objects
    const obj = { greeting: 'Hello {{name}}', data: { value: '{{count}}' } };
    const resolved = SDK.resolveObject(obj, context);
    console.log(`resolveObject(): ${JSON.stringify(resolved)}`);
}

// =============================================================================
// Main Runner
// =============================================================================
function main(): void {
    console.log('='.repeat(60));
    console.log('Runtime Template Resolver - Basic Usage Examples');
    console.log('='.repeat(60));

    example1_simpleResolution();
    example2_nestedPaths();
    example3_arrayAccess();
    example4_defaultValues();
    example5_missingStrategies();
    example6_resolveObject();
    example7_compilation();
    example8_extraction();
    example9_validation();
    example10_security();
    example11_sdkFunctions();

    console.log('\n' + '='.repeat(60));
    console.log('All examples completed!');
    console.log('='.repeat(60));
}

main();
